File: java/base-client/src/main/java/no/bankaxept/epayment/client/base/accesstoken/AccessToken.java
package no.bankaxept.epayment.client.base.accesstoken;

import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.regex.Pattern;

class AccessToken {
    private final String token;
    private final Instant expiry;

    public AccessToken(String token, long expiresIn, Clock clock) {
        this.token = token;
        this.expiry = clock.instant().plusSeconds(expiresIn);
    }

    public AccessToken(String token, long expiresIn) {
        this(token, expiresIn, Clock.systemDefaultZone());
    }

    static AccessToken parse(String input) {
        return Parser.parse(input);
    }

    static AccessToken parse(String input, Clock clock) {
        return Parser.parse(input, clock);
    }


    public String getToken() {
        return token;
    }

    public Instant getExpiry() {
        return expiry;
    }

    public long secondsUntilTenMinutesBeforeExpiry(Clock clock) {
        Instant tenMinutesBeforeExpiry = expiry.minus(10, ChronoUnit.MINUTES);
        return Duration.between(clock.instant(), tenMinutesBeforeExpiry).toSeconds();
    }

    private static class Parser {
        private static final Pattern tokenPattern = Pattern.compile("\"access_token\"\\s*:\\s*\"(.*)\"");
        private static final Pattern expiryPattern = Pattern.compile("\"expires_in\"\\s*:\\s*(\\d+)");

        static AccessToken parse(String input) {
            return parse(input, Clock.systemDefaultZone());
        }

        static AccessToken parse(String input, Clock clock) {
            var tokenMatcher = tokenPattern.matcher(input);
            var expiryMatcher = expiryPattern.matcher(input);
            if (!tokenMatcher.find() || !expiryMatcher.find()) {
                throw new IllegalArgumentException("Could not parse token or expiry");
            }
            return new AccessToken(tokenMatcher.group(1), Long.parseLong(expiryMatcher.group(1)), clock);
        }
    }
}

File: csharp/BaseClient/src/AccessToken/AccessToken.cs
using System;
using System.Text.RegularExpressions;

namespace BankAxept.Epayment.Client.Base.AccessToken {

    public class AccessToken {

        private readonly string token;
        private readonly DateTime expiry;
        
        public AccessToken(string token, long expiresIn, DateTime clock) {
            this.token = token;
            expiry = clock.AddSeconds(expiresIn);
        }

        public AccessToken(string token, long expiresIn) : this (token, expiresIn, DateTime.UtcNow) {}

        public string GetToken() {
            return token;
        }

        public DateTime GetExpiry() {
            return expiry;
        }

        public long SecondsUntilTenSecondsBeforeExpiry (DateTime clock) {
            return (long)(expiry.AddSeconds(-10) - clock).TotalSeconds;
        }

        public long SecondsUntilTenPercentBeforeExpiry (DateTime clock) {
            var remainingSeconds = (expiry - clock).TotalSeconds;
            return (long)(remainingSeconds * 0.9);
        }

        public static AccessToken Parse(string input) {
            return AccessTokenParser.Parse(input);
        }

        public static AccessToken Parse(string input, DateTime clock) {
            return AccessTokenParser.Parse(input, clock);
        }


        private static class AccessTokenParser {
            private static readonly Regex tokenPattern = new Regex("\"access_token\"\\s*:\\s*\"(.+?)\"");
            private static readonly Regex expiryPattern = new Regex("\"expires_in\"\\s*:\\s*(\\d+)");

            public static AccessToken Parse(string input) {
                return Parse(input, DateTime.UtcNow);
            }

            public static AccessToken Parse(string input, DateTime clock) {
                var tokenMatch = tokenPattern.Match(input);
                var expiryMatch = expiryPattern.Match(input);
                if (!tokenMatch.Success || !expiryMatch.Success) {
                    throw new ArgumentException("Could not parse token or expiry");
                }
                return new AccessToken(tokenMatch.Groups[1].Value, long.Parse(expiryMatch.Groups[1].Value), clock);
            }
        }
    }
}
